Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift

```diff
@@ -677,6 +676,13 @@ public final class SwiftTargetBuildDescription {
         self.buildParameters.triple.isDarwin() && self.target.type == .library
     }
 
+    private func writeFileList(includeDerivedSources: Bool = false) throws -> AbsolutePath {
```

Comment:
how are these files getting deleted?

Reply:
I think this function should take the array of sources instead of the includeDerivedSources argument, this leave control in the hand of the call site and not push the logic to this function

Reply:
I also hope if these files get deleted there was an option to not delete and keep them on the file system after the build for debugging purposes. Had exact same problem with `clang` today. Looks like that has a `-save-temps` flag, but apparently it respects that only for temporary `.o` files and similar temporary build artifacts. We don't have to have the same exact option, but some flag to not delete these files would be appreciated.

Reply:
The file is never deleted, it is emitted and kept, which I was happily surprised about.  If the files are reaped, it would mean that the command that is displayed is not possible to replicate as the file list is needed to execute the child process.

Sure, I like the idea of the taking the sources.

Reply:
@MaxDesiatov I had that same exact concern which is why I was happy that it doesn't get deleted 😄 That is exactly the reason the tests needed adjustment - the `.build` directories are now created early to store the file list.

Reply:
if they are not deleted, is there a risk of conflicts between runs? also, where exactly are they written to?

Reply:
I don't think that it is any higher risk than say running a build on the same tree in a VM and on the host.  They are written to `[target].build` (which is why the tests need to be adjusted - the directory is created eagerly as the file lists are written out just like the output file map and other such files).

Reply:
so this is in the `.build` directory, but what happens when the file already exit from previous runs?

Reply:
It will be over-written.  There is no risk of things getting mixed up AFAICT.  Do you have a concrete example of where you see this risk?

Reply:
i feel strongly that:
1. we should have a good debugging story
2. the code is easy to maintain

using files instead of list of source is hurting debugging so I would restrict that to when we _have_ to do it (eg only on Windows, or when we detect its needed), but if we feel we _have_ to do it, then I remain concerned about the files being left around and then getting mixed up in future invocation  (the side effect issue). if the code cleans it up I would feel more comfortable with this change, but that would kill debuggability, which why i am having a hard time with this change

maybe the code should make sure to delete these files before the invocation to eliminate the risk of mixups from future mistakes

Reply:
Can you explain _how_ this is hurting debugging?  I don't think that this actually hurts debugging in practice.  Deleting the file before invocation would not work - it would not be available when needed (when compiling the code).

Reply:
I think the real solution to that is having the file lists be generated during the build, then we would get deletion automatically through stale-file removal. That does seem separate from this PR, but we could do it as a pre-requisite.

Re: debugging, it seems like in `swift build --verbose`, we get both driver and individual frontend invocations. The frontend invocations are showing the concrete files verbatim even when using a file list, so that seems fine to me.

Reply:
the point I am trying to make is that we have a situation where code like `writeOutputFileMap` relies on side effects (files on disk). this is risky and I dont think we should add more of this kind of code (as this PR does), we should instead figure out a way to eliminate such side effect usage

obviously, we cannot simply delete these files at the end of the build because we need them to enable debugging, ie if you want to invoke the command manually

basically what I would like to see these files are purged before you invoke “swift build” again, to eliminate the side-effect risk

since this PR does not offer a way to deal with the side-effect risk, I was trying to come up with suggestions on how to use this only when necessary, so at least we reduce the chance of the risk manifesting itself. maybe @neonichu has ideas how to deal with it better

Reply:
The files are overwritten each time the build is planned, so it is in effect purging the file - it just overwrites it.  I would agree if the name of the file were not deterministic.

Reply:
We don't have a way to purge previous files outside of what is tracked by the build system, because we cannot know which files were exactly generated during another invocation. I don't think we should invent a new mechanism to do so, so the only viable solution seems to be to generate these files during the build which I believe was always the intent to eventually do, then the deletion comes for free from stale-file removal as I mentioned.

Reply:
@neonichu that captures exactly my sentiments as well.

Reply:
> don't think we should invent a new mechanism to do so, so the only viable solution seems to be to generate these files during the build which I believe was always the intent to eventually do, then the deletion comes for free from stale-file removal as I mentioned.

that sounds good to me

> To me that seems like something that is separate from this PR, though. I don't think we're increasing the risk of anything like this happening because already every target will have a link file list and every compilation command will have an output file map, so already every single command is subject to this.

still uncomfortable adding more fuel to the fire. how difficult is achieving the above ^^? if it is within reach maybe we can do that for `writeOutputFileMap` first then use that technique in this PR as well

Reply:
@neonichu how about we explore that direction a bit. if a heavy lift, we can merge and track as an open issue. otherwise, lets fix the architecture and not add more floors on shaky foundation

Reply:
@tomerd one thing to note - this is *blocking* SPM usage on windows.  We cannot build parts of the Swift project even (e.g. DocC).  Can you help get CMake support integrated into DocC in the mean time?

Reply:
@compnerd I am happy to take this change in its current form if it is put behind #if windows

Reply:
@tomerd how do we ensure that regressions are not introduced for that path?  Could we add an explicit manifest for the tests and run tests via CMake?

Reply:
not sure how this is different from the rest of SwiftPM in that regard

in any case, i think this is the best compromise we can for now, and we should def make it the default once we have a systematic way to handle the side effects which @neonichu is looking at

Reply:
I think it would be incorrect to put this change behind `#if windows` tbh, there's nothing Windows specific here. We can maybe live with it as a temporary thing, but it'll increase the code complexity significantly for no reason which seems worse to me than taking the theoretical risk of stale files.